### PR TITLE
Release Automation

### LINKED ISSUE #6228: Releases don't include architecture type
Looking at the releases, I can see no architecture type.

For instance, although I'm using Ubuntu 22, I won't be able to run the release because I'm on `arm64`.

Please release binaries for multiple architectures (including Linux `arm64` and macOS `arm64`).
 
image

### LINKED ISSUE #7776: Add github action to generate reproducible builds
Add a new GitHub Action to streamline our release process by generating reproducible builds via workflow.

- [ ] The action should be triggered on both tag push and workflow dispatch events.
- [ ] The action should produce builds for Fedora, Ubuntu Focal, Ubuntu Jammy, and Ubuntu Noble, along with a .zip file for upload on the release page.
- [ ] This will also require to integrate functionality from `tools/repro-build.sh`
- [ ] Remove unused Dockerfiles and scripts for a cleaner codebase. But do not delete the `build-release` and `repro-build` scripts as they are still essential for release verification and signing by other team members.

### PR DESCRIPTION
**Pull request by s373nZ**
This PR fixes #7776 by adding a new CI action called "Release" which is triggered when a release tag is pushed, or via manual dispatch. The action:

1. Does some sanity checking of the release tag against the branch and Git branch to short-circuit prior to lengthy build time.
2. Prepares the build environments, builds the release targets and uploads release artifacts in parallel, as a matrix strategy.
3. Merges the independent build artifacts into a final release artifact, deleting the intermediaries.
4. Signs the releases using a PGP key configured in the repository secret configuration.
5. Creates a new release in the repository and attaches the artifacts as assets.

Requires the addition of repository variables named `GPG_PRIVATE_KEY` and `GPG_PASSPHRASE` which represent a CI signing key. The release captain and signers would be expected to perform a local reproducible build and add their key to `SHA256SUMS.asc` as a manual process and re-upload that asset to the release.

Release documentation has been updated as a best guess for `rc1` instructions.

Fixes #7776
Fixes #6228

## Notes

- Modified the existing `tools/build-release.sh` script to allow for distribution-specific targets for the Ubuntu builds, for parallelization.
- Also added a `--without-zip` option to `tools/build-release.sh` script to skip creating a default archive when one just wants to sign the artifacts from Github.

## Testing

It's a little hard to test, as I think tag pushes and workflow_dispatches need to be in `master` to trigger. However, I have another branch with some testing configuration setup [here](https://github.com/s373nZ/lightning/tree/7776-release-builds-automation-test) and have been running the CI via a branch push (ex https://github.com/s373nZ/lightning/actions/runs/11728964685).

In order to test, one must:

1. Update the `CHANGELOG.md` file with a dummy release title line that will map to a release tag.
2. Tag the local branch with that release tag and push it.
3. Push the testing branch which is set to run by branch configuration.

See [this commit](https://github.com/ElementsProject/lightning/commit/9923c2e1053011ed70f2b8a1f4815894289f007b) for an example.

## Possible improvements or TODOs

- ~Audit and remove unused Dockerfiles and release tooling.~
- ~Add a PGP signing key to as a Github secret variable and do initial release signing as part of the CI as well.~
- ~Create the draft release object in CI and assign the release artifacts to it.~
- ~Try to address #6228 as well, by changing the release file names to include `amd64` in the Ubuntu target versions.~

## Checklist
Before submitting the PR, ensure the following tasks are completed. If an item is not applicable to your PR, please mark it as checked:

- [x] The changelog has been updated in the relevant commit(s) according to the [guidelines](https://docs.corelightning.org/docs/coding-style-guidelines#changelog-entries-in-commit-messages).
- [x] Tests have been added or modified to reflect the changes.
- [x] Documentation has been reviewed and updated as needed.
- [x] Related issues have been listed and linked, including any that this PR closes.

### CODE REVIEW

**Comment by s373nZ** [comment]
* Made a few tweaks to the shell script changes to satisfy `shellcheck`.
* Added a commit to remove the following seemingly old Dockerfiles:
    - action.yml
        - Seems to try and declare a reusable Github Action?
        - Reference to contrib/Dockerfile.tester is broken.
    - contrib/docker/Dockerfile.alpine
    - contrib/docker/Dockerfile.builder
    - contrib/docker/Dockerfile.tester
    - contrib/docker/Dockerfile.ubuntu
    - contrib/docker/scripts/*
    - I was tempted to move `contrib/docker/Dockerfile.builder.fedora` to `contrib/reprobuild/`, but it's technically non-reproducible. LMK if we think it should go there anyway.
* Changed the `tools/check-release.sh` to require a `--version` argument and changed the comments to make more sense after updating the script from reading `$GITHUB_*` variable to passing the version in as an argument.
* Removed an `echo` statement from debugging Ubuntu `$distribution` parsing.

**Comment by s373nZ** [comment]
Rebased against `master` and added a commit to change the Ubuntu release artifact filenames to include the architecture (`amd64`). Can move this to a different PR if you prefer.

Fixes #6228

**Comment by s373nZ** [comment]
One possible improvement, or point of discussion, would be to have manually triggered workflows just check out the tag that is specified by user input, instead of expecting the tag they enter to match that of HEAD.

Advantages are that one could manually build releases for arbitrary tags, avoids somewhat flaky expectation that tag == HEAD, and possibly more intuitive.

Disadvantage is that we might need duplicate `checkout` actions for each trigger, and less enforcement that releases are on the cutting edge of the `master` branch.

**Comment by s373nZ** [comment]
Rebased against `master`, fixed a typo in the documentation and updated the PR description with extra feature functionality and a note about the required secrets configuration for Github Actions.

**Comment by rustyrussell** [comment]
@ShahanaFarooqui wanted this in this release, at least for testing, so I've rebased.

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,66 @@
+---
+# https://docs.corelightning.org/docs/release-checklist
+name: "Release 🚀"
+on:
+  push:
+    tags:
+      - 'v[0-9]+.[0-9]+'
+      - 'v[0-9]+.[0-9]+.[0-9]+'
+      - 'v[0-9]+.[0-9]+[0-9a-z]+'
```
Comment (niftynei) [inline .github/workflows/release.yml:9]: what's an example of what this last regex is looking to match? 

it seems like it'd ignore things such as `v0.1.3-beta`. I believe the regex that you'd want for that would need to include the `.` and `-` in the options clause? Untested but...

`v[0-9]+.[0-9]+[0-9a-z.-]+`

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -13,7 +13,7 @@ jobs:
   check:
     name: Check
     outputs:
-      version: steps.capture.outputs.version
+      version: ${{ steps.capture.outputs.version }}
```
Comment (niftynei) [inline .github/workflows/release.yml:23]: nit: this change should be back-ported to the prior commit, where the error was first introduced

File: tools/repro-build.sh [hunk 3]
```diff
@@ -53,6 +53,7 @@ else
     VER=$(uname -r)
 fi
 
+ARCH=$(dpkg --print-architecture)
```
Comment (niftynei) [inline tools/repro-build.sh:56]: Unrelated nit, but iirc breaking your "changlog" notes into two lines will cause problems for the tool that auto reads them. eg 

```
Changelog-Changed: Added architecture identifier to Ubuntu release
artifacts. ([https://github.com/ElementsProject/lightning/issues/6228])
```

should be 

```
Changelog-Changed: Added architecture identifier to Ubuntu release artifacts. ([https://github.com/ElementsProject/lightning/issues/6228])
```

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -100,3 +100,52 @@ jobs:
           name: c-lightning-${{ env.version }}
           pattern: bin-*
           delete-merged: true
+
+  release:
```
Comment (niftynei) [inline .github/workflows/release.yml:111]: shouldn't this be called "sign_release"?

Comment (niftynei) [reply .github/workflows/release.yml:111]: ah i see you update it next :)

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -130,22 +137,45 @@ jobs:
           trust_level: 5
 
       - name: Set default GPG key
-        run: |
-          gpg -K
-          gpg --list-secret-keys
-          echo "default-key ${{ steps.gpg.outputs.keyid }}" >> ~/.gnupg/gpg.conf
-          cat ~/.gnupg/gpg.conf
+        run: echo "default-key ${{ steps.gpg.outputs.keyid }}" >> ~/.gnupg/gpg.conf
```
Comment (niftynei) [inline .github/workflows/release.yml:140]: is there a reason to not update the prior commit with this simplified method rather than editing it here?

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -130,22 +137,45 @@ jobs:
...
+          mv release/SHA256SUMS.asc${{ steps.gpg.outputs.keyid  }} release/SHA256SUMS.asc
 
       - name: Upload signed artifact
         uses: actions/upload-artifact@v4
         with:
           name: c-lightning-${{ env.version }}
           overwrite: true
           path: release/
+
+      - name: Determine release data
```
Comment (niftynei) [inline .github/workflows/release.yml:156]: nit: this should be in a separate commit from the previous changes in this file.

File: .github/workflows/release.yml (hunk 1 above)
Comment (rustyrussell) [reply .github/workflows/release.yml:9]: We literally append rcX, so "24.11rc3"...
